stdlib/public/TensorFlow/Ops.swift

```diff
@@ -1597,52 +1597,315 @@ public extension Tensor where Scalar : Numeric {
+    return slice(
+      lowerBounds: lowerBoundsTensor,
+      sizes: Tensor<Int32>(upperBounds) - lowerBoundsTensor)
+  }
+
+  @inlinable @inline(__always)
+  @differentiable(wrt: self, vjp: _vjpSlice)
+  func slice(lowerBounds: Tensor<Int32>, sizes: Tensor<Int32>) -> Tensor {
+    return Raw.slice(self, begin: lowerBounds, size: sizes)
+  }
+
+  @inlinable @inline(__always)
+  internal func _vjpSlice(
+    lowerBounds: Tensor<Int32>,
+    sizes: Tensor<Int32>
+  ) -> (Tensor, (Tensor) -> Tensor) {
+    let value = slice(lowerBounds: lowerBounds, sizes: sizes)
+    let afterPaddings = shapeTensor - value.shapeTensor - lowerBounds
+    return (value, { [after = afterPaddings] v in
+      let beforePaddings = lowerBounds.expandingShape(at: 1)
+      let afterPaddings = after.expandingShape(at: 1)
+      let paddings = Tensor<Int32>(
+        concatenating: [beforePaddings, afterPaddings], alongAxis: 1)
+      return Raw.pad(v, paddings: paddings)
+    })
+  }
+}
+
+public enum TensorRange : TensorRangeExpression {
+  case _ellipsis
+  case _newAxis
+  case _squeezeAxis
+  case _index(Int32)
+  case _range(Range<Int32>, stride: Int32)
+  case _closedRange(ClosedRange<Int32>, stride: Int32)
+  case _partialRangeFrom(PartialRangeFrom<Int32>, stride: Int32)
+  case _partialRangeUpTo(PartialRangeUpTo<Int32>, stride: Int32)
+  case _partialRangeThrough(PartialRangeThrough<Int32>, stride: Int32)
+
+  public var tensorRange: TensorRange { return self }
+}
+
+extension TensorRange : Equatable {
+  public static func == (lhs: TensorRange, rhs: TensorRange) -> Bool {
+    switch (lhs, rhs) {
+    case (._ellipsis, ._ellipsis): return true
+    case (._newAxis, ._newAxis): return true
+    case (._squeezeAxis, ._squeezeAxis): return true
+    case (let ._index(i1), let ._index(i2)): return i1 == i2
+    case (let ._range(r1, s1), let ._range(r2, s2)): return r1 == r2 && s1 == s2
+    case (let ._closedRange(r1, s1), let ._closedRange(r2, s2)):
+      return r1 == r2 && s1 == s2
+    case (let ._partialRangeFrom(r1, s1), let ._partialRangeFrom(r2, s2)):
+      return r1.lowerBound == r2.lowerBound && s1 == s2
+    case (let ._partialRangeUpTo(r1, s1), let ._partialRangeUpTo(r2, s2)):
+      return r1.upperBound == r2.upperBound && s1 == s2
+    case (let ._partialRangeThrough(r1, s1), let ._partialRangeThrough(r2, s2)):
+      return r1.upperBound == r2.upperBound && s1 == s2
+    default: return false
+    }
+  }
+}
+
+public protocol TensorRangeExpression {
+  var tensorRange: TensorRange { get }
+}
+
+public extension TensorRangeExpression {
+  static var ellipsis: TensorRangeExpression {
+    return TensorRange._ellipsis
+  }
+
+  static var newAxis: TensorRangeExpression {
+    return TensorRange._newAxis
+  }
+
+  static var squeezeAxis: TensorRangeExpression {
+    return TensorRange._squeezeAxis
+  }
+
+  static func index(_ value: Int32) -> TensorRangeExpression {
+    return TensorRange._index(value)
+  }
+
+  static func range(
```

Comment:
Do these static methods work in subscripts? It would be great if they are tested. They may not work in a straightforward way since they are defined in protocol extensions, not nominal type extensions.

Reply:
I hope to add tests for these tomorrow and see how to go ahead from there.

Reply:
@rxwei as you guessed, they do not work. Is there an established pattern for how to support them or should I just try to work around the limitation on my own?

Reply:
I'm thinking of using global variables and functions as an alternative.

Reply:
I implemented it using global variables and functions for now and added tests. This should be ready for review now. :)